Enhance Code Review guidelines (#1727)
uktrade/data-hub-leeloo: docs/Code review guidelines.md

- Follow Uncle Bob’s boy-scout rule: “always leave the code behind in a better state than you found it”
- Include unit, functional, e2e and compatibility tests (when applicable)
- Make sure CI build passes consistently without any flaky tests

#### Commit Hygiene
- Follow branch naming conventions by prefixing your branch name with *feature/*, *bugfix/*, *removal/*, *test/*, *hotfix/* or *release/*
- Make sure commits are logical and atomic - each commit should include tests
- Keep each commit small and deployable - each commit should ideally leave the develop branch in a releasable state
- Use [imperative mood](https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?id=HEAD#n133) in commit message but it’s okay to use past tense in the description

#### PR Hygiene
- Make sure your PR is atomic and doesn’t solve multiple problems at the same time
- Keep your PR small and deployable - each PR **must** leave the develop branch in a releasable state

> - The ideal maximum length for a PR is [400 lines of code](https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/)

Reviewer: I think there is a case for including a line like:
- The caveat here is to also bear in mind productivity. If there is a clear case for a larger PR that tackles more than one thing, e.g. when introducing brand new functionality, it is sensible to do that.

(I dialled into the meeting where this was discussed, but had to dial off before the end - so unsure whether this was discussed much)

Author: Yes, we discussed this. The overall impression was that a PR can always be split into smaller ones when adding brand new functionality. Having bigger PRs is counterproductive as they take more time to review and merge.

- Follow Uncle Bob's boy-scout rule: “always leave the code behind in a better state than you found it”
- Include unit, functional, e2e and compatibility tests (when applicable)
- Make sure CI build passes consistently without any flaky tests

#### Commit Hygiene
- Follow branch naming conventions by prefixing your branch name with *feature/*, *bugfix/*, *removal/*, *test/*, *hotfix/* or *release/*
- Make sure commits are logical and atomic - each commit should include tests
- Keep each commit small and deployable - each commit should ideally leave the develop branch in a releasable state
- Use [imperative mood](https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?id=HEAD#n133) in commit message but it's okay to use past tense in the description

#### PR Hygiene
- Make sure your PR is atomic and doesn't solve multiple problems at the same time
- Keep your PR small and deployable - each PR **must** leave the develop branch in a releasable state

> - The ideal maximum length for a PR is [400 lines of code](https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/)

Reviewer 2: I don't like the use of the word "ideal" here. On a first read it sounds like you should be aiming for that. "PRs should never normally be more than 400 lines of code, and are usually much less."?

Author: Changed it to _PRs shouldn't normally add or change more than 400 lines of code at the time_ after discussing it with other devs; hope it's okay!

- Include instructions on how to test (e.g. what should I see?)
- Detail any considerations when releasing

#### Screenshots
- Add before / after screenshots or GIFs
- Include screenshots of both mobile and desktop versions

## <a name="step-2"></a>Step 2: PR submitted

### For both authors and reviewers:

#### Attitude
- Remember that you are starting a conversation
- Don’t take feedback personally

> - Be honest but don't be rude

Reviewer: ```suggestion
- Be honest, but don't be rude
```

Reviewer 2: Don't give feedback personally. Be honest, but don't be rude.

#### GitHub
- Non-trivial questions are best asked via slack or face to face
- The same applies to PRs with large number of comments
- At the end of a conversation, update GitHub with a summary

### If you are the author:

- Make sure you read and follow the guidelines in [Step 1: before and when submitting a PR](#step-1)
- Don’t rebase or force-push whilst the PR is being reviewed otherwise reviewers won’t be able to see what’s changed
- Don’t dismiss change requests except in rare circumstances (e.g. when the reviewer is on holiday)
- Mark commits that will be squashed into other commits with the *fixup:* prefix. This helps reviewers understand how you are planning to organise your PR after approval
- Respond to comments if you don't agree with them

> - If you won’t implement the suggestion, come to a mutual understanding with the reviewer

Reviewer: ```suggestion
- If you decide not to implement the suggestion, come to a mutual understanding with the reviewer
```

### If you are a reviewer:

#### Time
- Allocate 30 minutes in the morning and 30 minutes in the afternoon for reviewing PRs

> - If there are multiple PRs, the preferred method for selection is FIFO (first in first out). However, this could result in fewer PRs being reviewed by multiple developers so use common sense

Reviewer: ```suggestion
- If there are multiple PRs, the preferred method for selection is FIFO (first in first out). However, this could result in fewer PRs being reviewed by multiple developers, so use common sense
```

- Whenever possible, review a PR in one go so that the author understands the amount of work needed and can plan with his/her team

#### Language
- Offer suggestions - “It might be easier to...”, “Consider...”
- Be objective - “this method is missing a docstring” instead of “you forgot to write a docstring”

#### Being diplomatic
- Feel free to constructively challenge approaches and solutions, even when coming from a seasoned developer

> - It’s OK to nit-pick including on syntax issues, spelling errors, poor variable/function names, missing corner cases

Reviewer: ```suggestion
- It’s OK to nit-pick on syntax issues, spelling errors, poor variable/function names, missing corner cases
```

> - Don’t be too much of a perfectionist - allow for some flexibility

Reviewer 3: Might be worth adding a 'But' to the beginning of this sente3nce to indicate it modifies the previous one

> - But don’t be too much of a perfectionist - allow for some flexibility

Reviewer: ```suggestion
- But don’t be a perfectionist - allow some flexibility
```

- Include instructions on how to test (e.g. what should I see?)
- Detail any considerations when releasing

#### Screenshots
- Add before / after screenshots or GIFs
- Include screenshots of both mobile and desktop versions

## <a name="step-2"></a>Step 2: PR submitted

### For both authors and reviewers:

#### Attitude
- Remember that you are starting a conversation
- Don’t take feedback personally
- Be honest, but don't be rude

#### GitHub

> - Non-trivial questions are best asked via slack or face to face

Reviewer 4: Capital `S`

- The same applies to PRs with large number of comments
- At the end of a conversation, update GitHub with a summary

### If you are the author:

- Make sure you read and follow the guidelines in [Step 1: before and when submitting a PR](#step-1)

> - Don’t rebase or force-push whilst the PR is being reviewed otherwise reviewers won’t be able to see what’s changed

Reviewer 4: Is could be worth writing "Don’t rebase or force-push" as "Don't `rebase` or `push --force`" to be clear about Git commands.

> - Don’t dismiss change requests except in rare circumstances (e.g. when the reviewer is on holiday)

Reviewer 4: Maybe worth adding "document the reason"

- Mark commits that will be squashed into other commits with the *fixup:* prefix. This helps reviewers understand how you are planning to organise your PR after approval
- Respond to comments if you don't agree with them
- If you decide not to implement the suggestion, come to a mutual understanding with the reviewer

### If you are a reviewer:

#### Time
- Allocate 30 minutes in the morning and 30 minutes in the afternoon for reviewing PRs
- If there are multiple PRs, the preferred method for selection is FIFO (first in first out). However, this could result in fewer PRs being reviewed by multiple developers, so use common sense
- Whenever possible, review a PR in one go so that the author understands the amount of work needed and can plan with his/her team

#### Language
- Offer suggestions - “It might be easier to...”, “Consider...”
- Be objective - “this method is missing a docstring” instead of “you forgot to write a docstring”

#### Being diplomatic
- Feel free to constructively challenge approaches and solutions, even when coming from a seasoned developer
- It's OK to nit-pick on syntax issues, spelling errors, poor variable/function names, missing corner cases

> - But don't be a perfectionist - allow some flexibility

Reviewer 2: Suggestion for additional point: it's probably better to talk any complex issues through face-to-face.

Author: 👍 it was already specified in an earlier point but I'll reword it

> #### Architectural feedback

Reviewer 2: I think this should be higher up.

- Focus on big architectural issues or problems with overall design first. If you spot any, give your feedback immediately before continuing with the review
- Check out the branch and run it locally for a broader view as GitHub tends to focus on single lines

#### Levels of importance
- Prefix non-critical comments with *Non-critical:* so that the author knows what is important and what is not
- If all your comments are non-critical, leave your feedback but accept the PR at the same time so that you are not a blocker

> 

Reviewer 2: Is it worth referrring back to "Attitude" above? The "remember you are starting a conversation" point is important.
